vsphere/datadog_checks/vsphere/api.py

```diff
@@ -0,0 +1,175 @@
+# (C) Datadog, Inc. 2019-present
+# All rights reserved
+# Licensed under Simplified BSD License (see LICENSE)
+import functools
+import ssl
+
+from pyVim import connect
+from pyVmomi import vim, vmodl
+
+from datadog_checks.base import ensure_unicode, is_affirmative
+from datadog_checks.vsphere.constants import (
+    ALL_RESOURCES,
+    DEFAULT_BATCH_COLLECTOR_SIZE,
+    MAX_QUERY_METRICS_OPTION,
+    UNLIMITED_HIST_METRICS_PER_QUERY,
+)
+
+
+def smart_retry(f):
+    """A function decorated with this `@smart_retry` will trigger a new authentication if it fails. The function
+    will then be retried.
+    This is useful when the integration keeps a semi-healthy connection to the vSphere API"""
+
+    @functools.wraps(f)
+    def wrapper(*args, **kwargs):
+        api_instance = args[0]
+        try:
+            return f(*args, **kwargs)
+        except Exception:
+            api_instance.smart_connect()
+            return f(*args, **kwargs)
+
+    return wrapper
+
+
+class APIConnectionError(Exception):
+    pass
+
+
+class VSphereAPI(object):
+    """Abstraction class over the vSphere SOAP api using the pyvmomi library"""
+
+    def __init__(self, instance):
+        self.host = instance['host']
+        self.username = instance['username']
+        self.password = instance['password']
+        self.ssl_verify = is_affirmative(instance.get('ssl_verify', True))
+        self.ssl_capath = instance.get('ssl_capath')
+        self.batch_collector_size = instance.get('batch_property_collector_size', DEFAULT_BATCH_COLLECTOR_SIZE)
+        self._conn = None
+        self.smart_connect()
+
+    def smart_connect(self):
+        """Creates the connection object to the vSphere API using parameters supplied from the configuration.
+        """
+        context = None
+        if not self.ssl_verify:
+            context = ssl.SSLContext(ssl.PROTOCOL_SSLv23)
+            context.verify_mode = ssl.CERT_NONE
+        elif self.ssl_capath:
+            context = ssl.SSLContext(ssl.PROTOCOL_SSLv23)
+            context.verify_mode = ssl.CERT_REQUIRED
+            context.load_verify_locations(capath=self.ssl_capath)
+
+        try:
+            # Object returned by SmartConnect is a ServerInstance
+            # https://www.vmware.com/support/developer/vc-sdk/visdk2xpubs/ReferenceGuide/vim.ServiceInstance.html
+            conn = connect.SmartConnect(host=self.host, user=self.username, pwd=self.password, sslContext=context)
+            conn.CurrentTime()
+        except Exception as e:
+            err_msg = "Connection to {} failed: {}".format(ensure_unicode(self.host), e)
+            raise APIConnectionError(err_msg)
```

Comment:
The exception message is logged. What else do you think would help? The username?

Reply:
If I understand correct, the `APIConnectionError` might caught by `smart_retry` decorator, but then it's not logged in any way.

Maybe we can log the exception caught in `smart_retry` ?

Reply:
No it will never be caught by `@smart_retry`,
1. the `smart_connect` method is not decorated with `@smart_retry`
2. The `smart_connect` method is only called in the `__init__` (not decorated) and inside the `smart_retry` decorator itself (but after the try/catch so any error will bubble up)

Reply:
Though it makes sense to `self.log.debug` in the `@smart_retry` decorator so that we don't ignore the exception completely.

Reply:
Yes, you are right, thx for the explanation.

> Though it makes sense to self.log.debug in the @smart_retry decorator so that we don't ignore the exception completely.

Yes, could help.